cli/klaus/src/main.rs

```diff
@@ -76,44 +76,50 @@ async fn main() -> Result<(), Error> {
     span.record("pid", args.pid);
     span.record("registry", traceable_path(&args.registry_path));
-    span.record("socket", maybe_traceable_path(&args.socket_path));
+    span.record("dot_flox_hash", &args.dot_flox_hash);
+    span.record("socket", traceable_path(&args.socket_path));
     span.record("log", maybe_traceable_path(&args.log_path));
-    if let Some(ref path) = args.socket_path {
-        debug!(socket_path = traceable_path(&path), "was provided a socket");
-    }
 
     debug!("starting");
+    debug!(
+        path = traceable_path(&args.socket_path),
+        exists = &args.socket_path.exists(),
+        "checked socket"
+    );
 
-    // The parent may have already died, in which case we just want to exit
-    if let Some(ref pid) = args.pid {
-        // TODO: re-use the method from ActivationPid after merging both PRs
-        let parent_is_running = match kill(Pid::from_raw(*pid), None) {
-            // These semantics come from kill(2).
-            Ok(_) => true,              // Process received the signal and is running.
-            Err(Errno::EPERM) => true,  // No permission to send a signal but we know it's running.
-            Err(Errno::ESRCH) => false, // No process running to receive the signal.
-            Err(_) => false,            // Unknown error, assume no running process.
-        };
-        if !parent_is_running {
-            return Err(anyhow!("detected that watchdog had unexpected parent"));
-        }
+    // Register activation PID so that we can track last one out
+    let activation = ActivationPid::from(args.pid);
+    if !activation.is_current_process_parent() {
+        return Err(anyhow!("detected that watchdog had unexpected parent"));
     }
+    register_activation(&args.registry_path, &args.dot_flox_hash, activation)?;
 
     // Start the listeners
     let shutdown_flag = Arc::new(AtomicBool::new(false));
     let signal_listener = signal_listener()?;
     let signal_task = spawn_signal_listener(signal_listener, shutdown_flag.clone())?;
-    let pid = target_pid(&args);
-    let termination_task = spawn_termination_listener(pid, shutdown_flag.clone());
+    let termination_task = spawn_termination_listener(activation.into(), shutdown_flag.clone());
 
     info!(
         this_pid = nix::unistd::getpid().as_raw(),
-        target_pid = pid.as_raw(),
+        target_pid = args.pid,
         "watchdog is on duty"
     );
 
     // Listen for a notification
-    let _action = listen(signal_task, termination_task, shutdown_flag).await;
+    let action = listen(signal_task, termination_task, shutdown_flag).await;
+
+    match action {
+        Action::Cleanup => {
+            info!(pid = &args.pid, "deregistering activation");
+            deregister_activation(&args.registry_path, &args.dot_flox_hash, activation)
+                .context("failed to deregister activation")?;
+        },
+        Action::Terminate => {
+            // TODO: deregister if !is_running?
```

Comment:
@zmitchell I seem to recall that you had a scenario in mind where we shouldn't do dereg on termination. Does that still apply if we know that the activation is no longer running?